Glass brain hemispheres

MartinPerez opened the pull request:
Glass brain hemispheres implementation to address #942.

tests, doc update and examples update included.

Here result images from the plot_demo_glass_brain_extensive example.

![figure_6](https://cloud.githubusercontent.com/assets/8766915/13400201/fce52e18-df06-11e5-9a84-4d4774788c0f.png)
![figure_7](https://cloud.githubusercontent.com/assets/8766915/13400202/fe471686-df06-11e5-95d7-eb4a92dbd866.png)
![figure_8](https://cloud.githubusercontent.com/assets/8766915/13400204/ffd9be68-df06-11e5-9f7b-682390d902a6.png)
![figure_9](https://cloud.githubusercontent.com/assets/8766915/13400205/01917494-df07-11e5-95e4-1ab45945a46b.png)
![figure_10](https://cloud.githubusercontent.com/assets/8766915/13400207/029bc1e6-df07-11e5-91d8-020fb114fb5c.png)

AlexandreAbraham commented:
This looks really nice!

MartinPerez commented:
sorry rebasing ...

GaelVaroquaux commented:
This does look very nice! On the images at the top (with a white background), there is a strange yellow background color.

MartinPerez commented:
You mean the two first images? I see it white :S

GaelVaroquaux commented:
I am positive that there is something in the background. I can see the rectangles.

MartinPerez commented:
in two screens (desktop, laptop) I see it white. No yellow background... @AlexandreAbraham what do you see?

AlexandreAbraham commented:
There is a yellow background. Color 255, 255, 251. Gimp can help:

![yellow](https://cloud.githubusercontent.com/assets/1647301/13401048/dd834a06-df0a-11e5-8603-331d16f13e8e.png)

GaelVaroquaux commented on examples/01_plotting/plot_demo_glass_brain.py:
```diff
@@ -20,10 +20,19 @@
+plotting.plot_glass_brain(
+    localizer_tmap_filename, display_mode='lyrz', threshold=3)
+
+plotting.plot_glass_brain(
```
We want to keep this example as simple as possible. I suggest that we remove that last line, which belongs more to the advanced example.

MartinPerez replied:
ok

GaelVaroquaux commented on examples/01_plotting/plot_demo_glass_brain_extensive.py:
```diff
@@ -35,4 +36,20 @@
 
+# Hemispheric saggital cuts
+plotting.plot_glass_brain(localizer_tmap_filename, threshold=3,
+                          display_mode='lyrz')
```
Can we have a demo of "lzr", which I suspect is the mode useful mode.

MartinPerez replied:
no problem

GaelVaroquaux commented on nilearn/plotting/displays.py:
```diff
@@ -350,6 +350,120 @@ def _add_lines(self, line_coords, line_values, cmap,
+
+    def draw_left_right(self, size, bg_color, **kwargs):
+        if self.direction in 'lr':
+            return
```
Am I wrong to think that if we modify the base class "draw_left_right" method, around line 115, we don't need to reimplement this method in the derived class?

GaelVaroquaux commented on nilearn/plotting/displays.py:
```diff
@@ -350,6 +350,120 @@ def _add_lines(self, line_coords, line_values, cmap,
+                bbox=dict(boxstyle="square,pad=0", ec=bg_color, fc=bg_color),
+                **kwargs)
+
+    def transform_to_2d(self, data, affine):
```
It seems to me that there's a lot of copy/paste with regards to the base-class method (in GlassBrainAxes). I would prefer modifying it to support l/r projection.

MartinPerez replied:
I also noticed that some changes would just mean one line in the parent classes. Nonetheless I put all the changes in the GlassBrainHemisphericAxes class for 2 reasons. First its not just GlassBrainAxes but also BaseAxes that would have modifications, I thought having spread changes across multiple classes for one feature might be confusing in the future and could break some other derived class (so far its assume for all the plotting framework that only xyz exist). Second I am not sure how the changes introduced would interact with the connectomics plots that also use the GlassBrainAxes class. I think putting this together and checking how this should be used with the connectomics should be a different PR. @GaelVaroquaux what do you think?

GaelVaroquaux replied:
I don't really mind if the modifications propagate up if it makes the code simpler. For classes that don't accept "l" or "r", we can always capture this in the __init__

GaelVaroquaux replied:
Also, I forgot to mention, but I would love the connectome plots to have logic for "l" and "r" too.

AlexandreAbraham replied:
+1 for pushing the code in upper classes. If you are afraid of breaking stuff, maybe we can do it in 2 times? Merging this one. and refactoring afterward.

MartinPerez replied:
I would really prefer to merge and refactor in another PR. Also I will need someone to give me a hand with the connectome since I dont work with these I am not sure I understand what it means for the connectome plots to have 'l' and 'r' logic. Would it be in the 'l' case just to overlay the same markers and lines flipped?

GaelVaroquaux commented on nilearn/plotting/displays.py:
```diff
@@ -1312,13 +1481,35 @@ class YZProjector(OrthoProjector):
+class RProjector(OrthoHemisphericProjector):
+    _cut_displayed = 'r'
+    _default_figsize = [2.6, 2.3]
+
```
I'd like an "lzr" mode, I think.

GaelVaroquaux commented on nilearn/plotting/img_plotting.py:
```diff
@@ -963,14 +963,16 @@ def plot_glass_brain(stat_map_img,
             The statistical map image. It needs to be in MNI space
             in order to align with the brain schematics.
-        output_file : string, or None, optional
+        output_file : string, or None, optional. 'ortho' by default.
```
Is this change intentional? I have the feeling that it is not right.

MartinPerez replied:
I just mention the default value already implemented. You mean I should not mention it, or that the default itself is wrong?

GaelVaroquaux replied:
> I just mention the default value already implemented. You mean I should not
> mention it, or that the default itself is wrong?

The default is "None": we are talking about "output_file", aren't we?

MartinPerez replied:
sorry you are right. mistake during the rebase merge just done.

GaelVaroquaux commented on nilearn/plotting/tests/test_displays.py:
```diff
@@ -40,6 +41,16 @@ def test_demo_ortho_projector():
+    with tempfile.TemporaryFile() as fp:
+        oprojector.savefig(fp)
+    oprojector.close()
+
```
Could we also test the functionality with a call to plot_glass_brain.

MartinPerez replied:
Like in the examples?

GaelVaroquaux replied:
Yes, with a simple smoketest.

MartinPerez commented:
I fixed the background issue and discovered on the way that I have some yellow daltonism, haha.

![figure_10](https://cloud.githubusercontent.com/assets/8766915/13402645/83d017bc-df11-11e5-9fab-4eb6c43bf6a6.png)

MartinPerez commented:
Addressed all comments. Added lyr, lzr and lzry and actually generalized the code to accept any order of labels in the future, which also simplified the hemispheric projector class. Added smoke test on test_img_plotting.py.

Examples here:

![figure_11](https://cloud.githubusercontent.com/assets/8766915/13405055/f95bfff4-df1c-11e5-947d-dd3a07e4d795.png)
![figure_12](https://cloud.githubusercontent.com/assets/8766915/13405057/fa809a7a-df1c-11e5-8e44-407ddc9d593e.png)

chrisfilo commented:
Excellent addition! LGTM.

MartinPerez commented:
circle ci seems to fail due to sphinx and doc generation? Can anyone please give me a hand to understand this error. Codacy fails for what was pointed out by @GaelVaroquaux  about code duplication in displays.py in GlassBrainHemisphericAxes class

GaelVaroquaux commented on nilearn/plotting/img_plotting.py:
```diff
@@ -963,14 +963,16 @@ def plot_glass_brain(stat_map_img,
             is saved to a file, and the display is closed.
-        display_mode: {'ortho', 'x', 'y', 'z', 'xy', 'xz', 'yz'}
+        display_mode : {'ortho', 'x', 'y', 'z', 'xz', 'yx', 'yz', 'l', 'r',
+                        'lr', 'lzr', 'lyr', 'lzry', 'lyrz'}. Default is 'ortho'.
```
Your problem with circleCI might be that you cannot have a line return here. It's possible that this just needs to be a very long line.

MartinPerez commented:
circle ci pass now. The appveyor new error seems unrelated to the changes of this PR.

======================================================================
ERROR: nilearn.decomposition.tests.test_base.test_mask_reducer
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Miniconda-x64\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\Miniconda-x64\lib\site-packages\nilearn\decomposition\tests\test_base.py", line 41, in test_mask_reducer
    n_jobs=2, random_state=0)
  File "C:\Miniconda-x64\lib\site-packages\nilearn\decomposition\base.py", line 110, in mask_and_reduce
    ) for img, confound in zip(imgs, confounds))
  File "C:\Miniconda-x64\lib\site-packages\sklearn\externals\joblib\parallel.py", line 818, in __call__
    self._terminate_pool()
  File "C:\Miniconda-x64\lib\site-packages\sklearn\externals\joblib\parallel.py", line 549, in _terminate_pool
    self._pool.terminate()  # terminate does a join()
  File "C:\Miniconda-x64\lib\site-packages\sklearn\externals\joblib\pool.py", line 583, in terminate
    super(MemmapingPool, self).terminate()
  File "C:\Miniconda-x64\lib\multiprocessing\pool.py", line 465, in terminate
    self._terminate()
  File "C:\Miniconda-x64\lib\multiprocessing\util.py", line 207, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "C:\Miniconda-x64\lib\multiprocessing\pool.py", line 513, in _terminate_pool
    p.terminate()
  File "C:\Miniconda-x64\lib\multiprocessing\process.py", line 137, in terminate
    self._popen.terminate()
  File "C:\Miniconda-x64\lib\multiprocessing\forking.py", line 312, in terminate
    _subprocess.TerminateProcess(int(self._handle), TERMINATE)
WindowsError: [Error 5] Access is denied

GaelVaroquaux commented on examples/01_plotting/plot_demo_glass_brain.py:
```diff
@@ -20,10 +20,15 @@
 # demo glass brain plotting
 from nilearn import plotting
 
+# Whole brain saggital cuts
```
To make the HTML view of the example simpler to comprehend, I think that it would be good to have 3 different "cells" in this example, one per plot.

GaelVaroquaux commented on nilearn/plotting/glass_brain.py:
```diff
@@ -87,12 +87,16 @@ def _get_json_and_transform(direction):
-                              'z': 'top'}
+                              'z': 'top',
+                              'l': 'left_side',
+                              'r': 'side'}
```
That last one should be "right_side", no?

MartinPerez replied:
r reuses the same image as x. So I left it with the same id when getting info from the files.

GaelVaroquaux replied:
> r reuses the same image as x. So I left it with the same id when getting info
> from the files.

Of course. Sorry for the noise.

chrisfilo commented:
BTW if you would like to check how well this performs on a diverse set of maps you can download this NeuroVault snapshot and generate figures for each map: https://purl.stanford.edu/qd938sk1220

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -350,6 +350,115 @@ def _add_lines(self, line_coords, line_values, cmap,
+
+        # set unselected brain hemisphere activations to 0
+        if self.direction == 'l':
+            data_selection = data.copy()
```
I'm pretty sure that this copy is not necessary. This is not a big deal, though.

MartinPerez replied:
How should it be done? I was just trying to avoid modifying the original data "passed by reference" to the transform_to_2d function. Could I instead just do data=data to create a new function local variable?

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -350,6 +350,115 @@ def _add_lines(self, line_coords, line_values, cmap,
+        # set unselected brain hemisphere activations to 0
+        if self.direction == 'l':
+            data_selection = data.copy()
+            data_selection[(data.shape[0]/2):, :, :] = 0
```
Assuming that data is centered seems dangerous to me. You should use the affine to get the "real" coordinate of the center.

MartinPerez replied:
I see your point, i will address this

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -350,6 +350,115 @@ def _add_lines(self, line_coords, line_values, cmap,
+        else:
+            maximum_intensity_data = np.abs(data_selection).max(axis=max_axis)
+
+        if self.direction == 'l':
```
Can you comment on these lines?

MartinPerez replied:
ok. enough to clarify i am flipping in the saggital cut?

MartinPerez commented:
@AlexandreAbraham I addressed your comments. We merge the PR and start a new one to push to upper classes and add connectomics?

MartinPerez commented:
@chrisfilo thanks for the reference, I will check it out.

banilo commented on nilearn/plotting/displays.py:
```diff
@@ -350,6 +350,121 @@ def _add_lines(self, line_coords, line_values, cmap,
+    """An MPL axis-like object that displays a 2D projection of 3D
+    volumes with a schematic view of the brain. In the case of saggital cuts,
+    projects only data in the selected left or right hemisphere.
+
```
empty line

banilo commented on nilearn/plotting/displays.py:
```diff
@@ -350,6 +350,121 @@ def _add_lines(self, line_coords, line_values, cmap,
+
+    def draw_2d(self, data_2d, data_bounds, bounding_box,
+                type='imshow', **kwargs):
+        # kwargs massaging
```
massage or message? :-P

GaelVaroquaux commented:
> We merge the PR and start a new one to push to upper classes and add connectomics?

I would like the code duplication to be addressed before this lands in
master. Code duplication makes the repo hard to read and understand. I
don't want to degrade the quality of the repo, even with the promise that
it will improve later.

MartinPerez commented:
> I would like the code duplication to be addressed before this lands in master
ok I will work on the upper classes then. 

Who can give me a hand with checking the logic of the connectomics, or that I leave for later?

MartinPerez commented:
![connectome](https://cloud.githubusercontent.com/assets/8766915/13916520/737e06e4-ef5b-11e5-8ae5-b791339be912.png)

well connectome included and code generalized to upper classes. I also figured out that it was completely unnecessary to have the svg image for the right hemisphere, since the matplotlib x axis could be inverted.

MartinPerez commented:
of course added smoke test and example update for connectome

aabadie commented on examples/01_plotting/plot_demo_glass_brain.py:
```diff
@@ -26,4 +26,11 @@
     localizer_tmap_filename, title='plot_glass_brain',
     black_bg=True, display_mode='xz', threshold=3)
 
+# plotting.show()
```
Maybe this comment could be removed.

aabadie commented on examples/01_plotting/plot_demo_glass_brain.py:
```diff
@@ -17,7 +17,7 @@
 
 ###############################################################################
-# demo glass brain plotting
+# demo glass brain plotting. Whole brain saggital cuts
```
"Demo glass brain plotting using whole brain saggital cuts."

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -237,7 +245,24 @@ def transform_to_2d(self, data, affine):
+            data_selection = data.copy()
+            x_center, _, _, _ = np.dot(np.linalg.inv(affine),
+                                       np.array([0, 0, 0, 1]))
+            data_selection[int(x_center):, :, :] = 0
```
This will not work if the affine is not diagonal. Is that always the case? If so, you could just do `data_selection=data[:int(x_center), :, :]` to avoid the copy above.

MartinPerez replied:
@AlexandreAbraham as I understood from http://nipy.org/nibabel/coordinate_systems.html, the affine is not necessarily diagonal but is invertible and its last row is 0,0,0,1. Nibabel documentation explain that [i,j,k,1] = dot(inv(M), [x,y,z,1]) so I am trusting thats the case. Since for the glass brain we are in MNI space I am assuming I can consider space coordinates 0,0,0 to ask for the actual voxel indices corresponding to the center.

AlexandreAbraham commented:
I just have one question. When addressed (along with the remaining comment of aabadie), I will merge.

MartinPerez commented on nilearn/plotting/displays.py:
```diff
@@ -237,7 +245,27 @@ def transform_to_2d(self, data, affine):
+
+        # We need to make sure data_selection is not empty in the x axis
+        if data_selection.shape[0] == 0:
+            data_selection = data
```
I am not sure this is a good way to avoid a crash during smoke testing. Basically this will avoid the function crashing if the img passed has an affine that maps coordinate 0mm to a voxel index out of the array. How should handle that case. Actually raise a warning/error to the user?

MartinPerez commented:
@AlexandreAbraham Its fine now, but there is a small problem if the user tried to pass an img not in mni space, which should not be the case anyway. Could you please give me your opinion on how to handle that?

KamalakerDadi commented:
> but there is a small problem if the user tried to pass an img not in mni space, which should not be the case anyway. Could you please give me your opinion on how to handle that?

For me solution to this problem is to add a line saying that images should be in MNI space. We always pressure this point to users to avoid surprising results.

GaelVaroquaux commented:
> For me solution to this problem is to add a line saying that images should be
> in MNI space. We always pressure this point to users to avoid surprising
> results.

+1

KamalakerDadi commented:
Please also add a line about this new feature in ```whats_new.rst```

MartinPerez commented:
@KamalakerDadi I created version 0.2.5 in whats_new.rst, I assume thats the next release.

MartinPerez commented:
@AlexandreAbraham it seems there is an unrelated error in travis.
ERROR: nilearn.tests.test_testing.test_memory_usage

KamalakerDadi commented on examples/01_plotting/plot_demo_glass_brain.py:
```diff
@@ -17,7 +17,7 @@
 
 ###############################################################################
-# demo glass brain plotting
+# Demo glass brain plotting using whole brain saggital cuts
```
what it means by whole brain saggital cuts ?

KamalakerDadi replied:
saggital -> sagittal

KamalakerDadi commented on examples/01_plotting/plot_demo_glass_brain.py:
```diff
@@ -26,4 +26,9 @@
+###############################################################################
+# Hemispheric saggital cuts
+plotting.plot_glass_brain(
+    localizer_tmap_filename, display_mode='lyrz', threshold=3)
```
It would be easy to catch quickly using title something like plot_glass_brain with display_mode='lyrz'

KamalakerDadi commented on examples/01_plotting/plot_demo_glass_brain_extensive.py:
```diff
@@ -36,3 +37,25 @@
+###############################################################################
+# Hemispheric saggital cuts
+plotting.plot_glass_brain(localizer_tmap_filename, threshold=3,
+                          display_mode='lyrz')
```
same with here some ```title```s would be useful.

MartinPerez commented:
took care of all comments and tests passed except for the unrelated memory profiler one in travis.

lesteve commented on doc/whats_new.rst:
```diff
@@ -1,3 +1,14 @@
+---------
+
+New features
+
```
This needs to be a header, look at how it is done for the 0.2.4 section.

lesteve commented on examples/01_plotting/plot_demo_glass_brain_extensive.py:
```diff
@@ -36,3 +37,31 @@
+plotting.plot_glass_brain(localizer_tmap_filename, threshold=3, title=title,
+                          colorbar=True, plot_abs=False, display_mode='lyrz')
+
+plotting.show()
```
Dow we really need all this variations?

AlexandreAbraham replied:
+1, I guess 2 plots are enough ;).

lesteve commented on nilearn/plotting/displays.py:
```diff
@@ -266,9 +295,23 @@ def draw_position(self, size, bg_color, **kwargs):
+            xcoords, ycoords, zcoords = marker_coords.T
+            for cidx, xc in enumerate(xcoords):
+                if self.direction == 'r':
+                    if xc >= 0:
```
you can put that in the same if clause: `if self.direction == 'r' and xc >=0`

lesteve commented on nilearn/plotting/displays.py:
```diff
@@ -266,9 +295,23 @@ def draw_position(self, size, bg_color, **kwargs):
+                    if xc >= 0:
+                        relevant_coords.append(cidx)
+                elif self.direction == 'l':
+                    if xc < 0:
```
same remark as above you can put it in the same if clause.

lesteve replied:
By the way just wondering what should we do if xc == 0 (maybe that never happens in practice I am not sure). Should we not show the marker on both 'l' and 'r' plots?

GaelVaroquaux replied:
> By the way just wondering what should we do if xc == 0 (maybe that never
> happens in practice I am not sure). Should we not show the marker on both 'l'
> and 'r' plots?

Actually, that's probably a good idea. And that way we have edges going
to it in both plot (this needs to be documented in the docstrings of
add_marker and plot_connectome).

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -957,20 +957,26 @@ def plot_glass_brain(stat_map_img,
             The statistical map image. It needs to be in MNI space
             in order to align with the brain schematics.
-        output_file : string, or None, optional
+        output_file : string, or None, optional.
```
I wouldn't add a dot at the end since that's not what is done for other parameters

lesteve commented:
It would be worth to modify the docs to show the new modes in http://nilearn.github.io/plotting/index.html#different-display-modes.

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -237,7 +245,28 @@ def transform_to_2d(self, data, affine):
+                                       np.array([0, 0, 0, 1]))
+            data_selection = data[:int(x_center), :, :]
+        else:
+            data_selection = data
```
The code above will only work if the affine is diagonal. Are we sure that it is diagonal at this point?

MartinPerez replied:
@AlexandreAbraham as I understood from http://nipy.org/nibabel/coordinate_systems.html, the affine is not necessarily diagonal but is invertible and its last row is 0,0,0,1. Nibabel documentation explain that [i,j,k,1] = dot(inv(M), [x,y,z,1]) so I am trusting thats the case. Since for the glass brain we are in MNI space I am assuming I can consider space coordinates 0,0,0 to ask for the actual voxel indices corresponding to the center.

Does this answer your question?

GaelVaroquaux commented:
What's the status on the hemisphere plotting of connectomes in this PR?

PS: sorry, I think that I just introduced a trivial conflict with master.

MartinPerez commented:
@GaelVaroquaux its also ready for the connectome. Look for the image above.

GaelVaroquaux commented:
Indeed, very nice!

We would need a tiny example demoing it.

Maybe you can do a new example in the plotting section that demos plot connectome and replace the image in https://circle-artifacts.com/gh/nilearn/nilearn/1495/artifacts/0/home/ubuntu/nilearn/doc/_build/html/plotting/index.html#different-plotting-functions by the image it generates (and link to that example).

I think that, for now, we could use a hard-coded set of coordinates and connectivity matrix, for instance inspired by: https://circle-artifacts.com/gh/nilearn/nilearn/1495/artifacts/0/home/ubuntu/nilearn/doc/_build/html/auto_examples/03_connectivity/plot_adhd_spheres.html#sphx-glr-auto-examples-03-connectivity-plot-adhd-spheres-py

It would enable us to demo the hemisphere plotting for connectomes.

KamalakerDadi commented:
>Should I change the folder name or is this a new standard to enforce in all links?

No, you don't have to it will generate it automatically while you ```make html``` in doc folder.

MartinPerez commented:
@lesteve I illustrate the new display modes using two examples from glass brain. Did not make it extensive in the images but instead added reference to the remaining possible modes in the text at the side.

@GaelVaroquaux I modified the simple example you mentioned and emphasized the case x==0.
https://circle-artifacts.com/gh/nilearn/nilearn/1520/artifacts/0/home/ubuntu/nilearn/doc/_build/html/auto_examples/03_connectivity/plot_adhd_spheres.html

Also made link to image and examples in https://circle-artifacts.com/gh/nilearn/nilearn/1520/artifacts/0/home/ubuntu/nilearn/doc/_build/html/plotting/index.html#different-display-modes

Also addressed remaining comments.

bthirion commented:
LGTM.

AlexandreAbraham commented:
LGTM too. Merging. Thanks a lot!

GaelVaroquaux commented:
Whoooot

MartinPerez commented:
Awesome! thanks!
On Jun 7, 2016 5:11 PM, "Gael Varoquaux" <notifications@github.com> wrote:

> Whoooot
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nilearn/nilearn/pull/1034#issuecomment-224311972>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AIXFw3NsUXQYwkzt_0SRwEMZroAcDQhBks5qJYobgaJpZM4HljwT>
> .
>
